jscomp/syntax/src/res_core.ml

```diff
@@ -2260,56 +2331,33 @@ and parseTemplateExpr ?(prefix = "js") p =
     Ast_helper.Exp.apply ~attrs:[templateLiteralAttr] ~loc hiddenOperator
       [(Nolabel, e1); (Nolabel, e2)]
   in
-  let rec parseParts (acc : Parsetree.expression) =
-    let startPos = p.Parser.startPos in
-    Parser.nextTemplateLiteralToken p;
-    match p.token with
-    | TemplateTail (txt, lastPos) ->
-      Parser.next p;
-      let loc = mkLoc startPos lastPos in
-      let str =
-        Ast_helper.Exp.constant ~attrs:[templateLiteralAttr] ~loc
-          (Pconst_string (txt, Some prefix))
-      in
-      concat acc str
-    | TemplatePart (txt, lastPos) ->
-      Parser.next p;
-      let loc = mkLoc startPos lastPos in
-      let expr = parseExprBlock p in
-      let str =
-        Ast_helper.Exp.constant ~attrs:[templateLiteralAttr] ~loc
-          (Pconst_string (txt, Some prefix))
-      in
-      let next =
-        let a = concat acc str in
-        concat a expr
-      in
-      parseParts next
-    | token ->
-      Parser.err p (Diagnostics.unexpected token p.breadcrumbs);
-      Ast_helper.Exp.constant (Pconst_string ("", None))
-  in
-  let startPos = p.startPos in
-  Parser.nextTemplateLiteralToken p;
-  match p.token with
-  | TemplateTail (txt, lastPos) ->
-    Parser.next p;
-    Ast_helper.Exp.constant ~attrs:[templateLiteralAttr]
-      ~loc:(mkLoc startPos lastPos)
-      (Pconst_string (txt, Some prefix))
-  | TemplatePart (txt, lastPos) ->
-    Parser.next p;
-    let constantLoc = mkLoc startPos lastPos in
-    let expr = parseExprBlock p in
-    let str =
-      Ast_helper.Exp.constant ~attrs:[templateLiteralAttr] ~loc:constantLoc
-        (Pconst_string (txt, Some prefix))
+  let genInterpolatedString () =
+    let subparts =
+      List.flatten
+        (List.map
+           (fun part ->
+             match part with
+             | s, Some v -> [s; v]
+             | s, None -> [s])
+           parts)
     in
-    let next = concat str expr in
-    parseParts next
-  | token ->
-    Parser.err p (Diagnostics.unexpected token p.breadcrumbs);
-    Ast_helper.Exp.constant (Pconst_string ("", None))
+    let exprOption =
+      List.fold_left
+        (fun acc subpart ->
+          Some
+            (match acc with
+            | Some expr -> concat expr subpart
+            | None -> subpart))
+        None subparts
+    in
+    match exprOption with
+    | Some expr -> expr
+    | None -> Ast_helper.Exp.constant (Pconst_string ("", None))
+  in
+
+  match prefix with
+  | "js" | "j" | "json" -> genInterpolatedString ()
```

Comment:
Can you explain where the need to support `json` is coming from.
Not sure if it's something you want to support, or if it's here just to fix some test?

Reply:
The need is not really to support `json` but to avoid considering ``` `json`something`  ``` as a tagged template.

But now that js and j string interpolations have been simplified, we can likely refactor this function to something simpler.

My goal was to first make the original PR work, then start polishing it.

Reply:
Sounds good. The reason for asking is I would like to eventually remove that json thing entirely from externals too.